adl/rot13adl/rot13node_test.go

```diff
@@ -0,0 +1,78 @@
+package rot13adl
+
+import (
+	"testing"
+
+	. "github.com/warpfork/go-wish"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/node/basic"
+)
+
+func TestLogicalNodeRoundtrip(t *testing.T) {
+	// Build high level node.
+	nb := Prototype.Node.NewBuilder()
+	err := nb.AssignString("abcd")
+	Require(t, err, ShouldEqual, nil)
+	n := nb.Build()
+	// Inspect the high level node.
+	s, err := n.AsString()
+	Wish(t, err, ShouldEqual, nil)
+	Wish(t, s, ShouldEqual, "abcd")
+}
+
+func TestNodeInternal(t *testing.T) {
+	// Build high level node.
+	nb := Prototype.Node.NewBuilder()
+	err := nb.AssignString("abcd")
+	Require(t, err, ShouldEqual, nil)
+	n := nb.Build()
+	// Poke its insides directly to see that the transformation occured.
+	Wish(t, n.(*_R13String).synthesized, ShouldEqual, "abcd")
+	Wish(t, n.(*_R13String).raw, ShouldEqual, "nopq")
+}
+
+func TestReify(t *testing.T) {
+	t.Run("using unspecialized raw node", func(t *testing.T) {
+		// Build substrate-shaped data using basicnode.
+		sn := basicnode.NewString("nopq")
+		// Reify it.
+		synth, err := Reify(sn)
+		// Inspect the resulting high level node.
+		Require(t, err, ShouldEqual, nil)
+		Wish(t, synth.ReprKind(), ShouldEqual, ipld.ReprKind_String)
+		s, err := synth.AsString()
+		Wish(t, err, ShouldEqual, nil)
+		Wish(t, s, ShouldEqual, "abcd")
+	})
+	t.Run("using substrate node", func(t *testing.T) {
+		// Build substrate-shaped data, in the substrate type right from the start.
+		snb := Prototype.SubstrateRoot.NewBuilder()
+		snb.AssignString("nopq")
+		sn := snb.Build()
+		// Reify it.
+		synth, err := Reify(sn)
+		// Inspect the resulting high level node.
+		Require(t, err, ShouldEqual, nil)
+		Wish(t, synth.ReprKind(), ShouldEqual, ipld.ReprKind_String)
+		s, err := synth.AsString()
+		Wish(t, err, ShouldEqual, nil)
+		Wish(t, s, ShouldEqual, "abcd")
+	})
+}
+
+func TestInspectingSubstrate(t *testing.T) {
+	// Build high level node.
+	nb := Prototype.Node.NewBuilder()
+	err := nb.AssignString("abcd")
+	Require(t, err, ShouldEqual, nil)
+	n := nb.Build()
+	// Ask it about its substrate, and inspect that.
+	sn := n.(*_R13String).Substrate()
+	// TODO: the cast above isn't available outside this package: we should probably make an interface with `Substrate()` and make it available.
+	//  Is it reasonable to make this part of a standard feature detection pattern,
+	//   and make that interface reside in the main IPLD package?  Or in an `adl` package that contains such standard interfaces?
```

Comment:
Here's another place where it seems like a consistent interface for detecting ADLs might be useful.  What do with think of this?  yes/no/maybe?

Reply:
I'm gonna go with "maybe", because I agree that this should be consistent and easy between ADLs, but I don't like the name "substrate" :) 

How about `UnderlyingNode`? Or if you want to clarify that it's just for ADLs, perhaps something like `UnderlyingDataLayout`.

Reply:
So I have to confess this particular suggestion of alternative name made me laugh because of personal history.

(For whatever reason, in some of my earliest programming projects, I *really* latched onto the word "underlying".  I overused it horribly, everywhere.  At some point I became actively self-conscious of it and started just stripping the word out of the sentence every time I was tempted to use it, just like I try to strip "simply" nowadays.)

I don't want to say it's an objectively bad word, because it isn't, but this just gave me a hoot.

Reply:
A little more justifiably: I think I prefer "substrate" to "underlying" because "substrate" sounds a bit more like it's a precise term-of-art, and I think it'll be clearer to write docs and specs which strongly associate that word to ADLs.  By contrast, "underlying" is an adjective that could be applied to several parts of IPLD: e.g., one could also say "typed nodes have both a typed interface and an _underlying_ representation interface" and it would seem to make sense; this means it's a little harder to bake that word into a term-of-art.

Reply:
That sounds fine, but then we have to define "substrate" in the specs/docs early on, which I don't think we do right now.